pkg/gardenlet/controller/shoot/shoot_control_reconcile.go

```diff
@@ -428,6 +428,16 @@ func (r *shootReconciler) runReconcileShootFlow(ctx context.Context, o *operatio
 			}).RetryUntilTimeout(defaultInterval, defaultTimeout).SkipIf(o.Shoot.HibernationEnabled),
 			Dependencies: flow.NewTaskIDs(deployGardenerResourceManager, initializeShootClients, waitUntilOperatingSystemConfigReady, deployKubeScheduler),
 		})
+		_ = g.Add(flow.Task{
+			Name:         "Destroy node-local-dns system component",
+			Fn:           flow.TaskFn(botanist.DestroyNodeLocalDNS).SkipIf(o.Shoot.HibernationEnabled || o.Shoot.NodeLocalDNSEnabled),
+			Dependencies: flow.NewTaskIDs(deployGardenerResourceManager, initializeShootClients, waitUntilOperatingSystemConfigReady, deployKubeScheduler),
+		})
+		_ = g.Add(flow.Task{
+			Name:         "Deploying node-local-dns system component",
+			Fn:           flow.TaskFn(botanist.DeployNodeLocalDNS).RetryUntilTimeout(defaultInterval, defaultTimeout).SkipIf(o.Shoot.HibernationEnabled || !o.Shoot.NodeLocalDNSEnabled),
+			Dependencies: flow.NewTaskIDs(deployGardenerResourceManager, initializeShootClients, waitUntilOperatingSystemConfigReady, deployKubeScheduler),
+		})
```

Comment:
Can we merge this together to declutter it? Maybe have a function `ReconcileNodeLocalDNS` in the `botanist` which does either call `Deploy` or `Destroy` depending on `o.Shoot.NodeLocalDNSEnabled`. Then you just need to add one new task here.